core/include/clusterization/component_connection.hpp

```diff
@@ -46,32 +44,11 @@ class component_connection
     /// c++20 piping interface:
     /// @return a cluster collection
     ///
-    cluster_collection operator()(
-        const std::pair<const host_cell_collection&, const cell_module&>& i)
-        const override {
-        const host_cell_collection& cells = i.first;
-        const cell_module& module = i.second;
+    cluster_collection operator()(host_cell_collection&& cells,
```

Comment:
😕 Is it really a good idea to force the user to hand over their collection to the algorithm? What if they would like to run multiple algorithms on the same `host_cell_collection`?

I just don't see why the previous `const host_cell_collection&` argument is not better than moving the object into the operator... Did you want to make sure that the object would be deleted as early as possible? But again, what if the user wants to use that collection after having run the algorithm?

Reply:
I think this is a fair point; it is true that this kind of formalism can potentially get in the way if you're thinking about data re-use, but I would argue that traccc is a great place to try to incorporate affine typing (move semantics-ish) into our code. This, for me, is partly based in practical considerations, but also partially in more long-term programming language design.

traccc, in my view, is inherently a pipeline project, where we take data about pixel activations and turn them into tracks. At each step, we transform the data in some way, and we move forward through our chain of algorithms. It is a use case like this were move semantics are particularly powerful, since they allow us to quite literally model that movement of data. Do we really need to keep track of all of our old data all the time? Perhaps in some cases that is valuable, but surely in most cases we simply want to move on to the next algorithm in our chain, right? In cases where we really do need that old data, we can always make a copy of it, but I don't think that storing a copy should be the default approach.

Move semantics also serve a particularly useful purpose in high-memory usage heterogeneous computing, because they can guarantee at a type level (although the implementation is still somewhat iffy sometimes) that we can re-use memory that we no longer need. If you program your heterogeneous code cleverly that allows you to avoid a lot of allocations, since you can simply write your output to where your input used to be! You and I have talked about asynchronous allocation as well recently, and that becomes significantly easier with move semantics, too, since you can delegate the management of your memory to the kernel which is already running asynchronously. If you use move semantics properly you can get away with using less memory for your kernels, or conversely with running larger payloads on the same GPU!

Substructural typing is something we should not be ignoring, especially working on the border of computing and physics. I genuinely believe that more advanced type systems (especially affine and linear ones) which are now mostly academic toys will explode in popularity in the near future, if you consider for example how neatly the no-cloning theorem of quantum computing maps onto these systems! Of course modern C++ is not close to capturing a rich enough type theory to accomplish this, but we should not blindly follow the tried-and-tested "pass a const reference to everything" strategy, if we don't have to!

That said, I am a little bit of a dreamer in these regards. If you think passing these as const references for now I am more than happy to change the code to do that instead. :smile: